### PR TITLE
Use socket2 crate instead of reimplementing socket options.

### PR DESCRIPTION
**Pull request by PiotrSikora**
While there, fix original destination detection for IPv6 traffic.

Signed-off-by: Piotr Sikora <piotr@aviatrix.com>

### CODE REVIEW

**Comment by PiotrSikora** [comment]
Significant portion of this PR is pending review at https://github.com/rust-lang/socket2/pull/360, but I've decided to also copy it here, since it feels much safer than the code we're using right now.

**Review by hzxuzhonghu — APPROVED** [review]

File: src/socket.rs [hunk 1]
```diff
@@ -73,99 +68,51 @@ pub fn set_transparent(_: &TcpListener) -> io::Result<()> {
...
-            let e = io::Error::last_os_error();
-            warn!("failed to read SO_ORIGINAL_DST: {:?}", e);
-            return Err(e);
+    use std::os::unix::io::AsRawFd;
+
+    use socket2::{SockAddr, SockRef};
+    use tokio::io;
+
+    // Replace with socket2's version once there is a release that contains
+    // https://github.com/rust-lang/socket2/pull/360
```
Comment (howardjohn) [inline src/socket.rs:77]: this is merged. I guess we may need to wait for a release though?

**Review by howardjohn — APPROVED** [review]

File: src/socket.rs (hunk 1 above)
Comment (PiotrSikora) [reply src/socket.rs:77]: Yeah, especially because `master` started development for `v0.5.x` and they didn't make any releases of that yet.

**Pull request merged**